source/MRMesh/MRPointsComponents.cpp

```diff
@@ -0,0 +1,76 @@
+#include "MRPointsComponents.h"
+#include "MRPointCloud.h"
+#include "MRBitSet.h"
+#include "MRTimer.h"
+#include "MRPointsInBall.h"
+
+namespace MR
+{
+
+namespace PointCloudComponents
+{
+
+std::vector<MR::VertBitSet> getAllComponents( const PointCloud& pointCloud, float maxDist )
+{
+    MR_TIMER
+
+    assert( maxDist > 0.f );
+    const auto& validPoints = pointCloud.validPoints;
+    auto unionSstructs = PointCloudComponents::getUnionFindStructureVerts( pointCloud, maxDist );
+    auto allRoots = unionSstructs.roots();
+    std::vector<VertBitSet> components;
+    std::vector<VertId> componentRoots;
+
+    for ( auto v : validPoints )
+    {
+        auto componentId = -1;
+        VertId root = allRoots[v];
+        for ( int i = 0; i < componentRoots.size(); ++i )
+            if ( componentRoots[i] == root )
+            {
+                componentId = i;
+                break;
+            }
```

Comment:
It is too slow to search that way. I think in MeshComponents we use something like Hash for that.